### PR TITLE
Avoid raising the sampler warning if num_replicas=1

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Dont' raise a warning if devices=1 already and ddp is still used.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @ninginthecloud @rohitgr7 @otaj @akihironitta

### CODE REVIEW

**Comment by awaelchli** [comment]
@rohitgr7 We used to fall back to single device when `strategy="ddp"` and `devices=1`. We wouldn't even cross this code path since no distributed sampler is required in this case. This must have been lost in the accelerator connector "refactor".

File: src/pytorch_lightning/trainer/connectors/data_connector.py [hunk 1]
```diff
@@ -298,12 +298,16 @@ def _resolve_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional
...
+                and trainer_fn in (TrainerFn.VALIDATING, TrainerFn.TESTING)
+            ):
                 rank_zero_warn(
-                    f"Using `DistributedSampler` with the dataloaders. During `trainer.{trainer_fn.value}()`,"
-                    " it is recommended to use `Trainer(devices=1)` to ensure each sample/batch gets evaluated"
+                    f"Using `DistributedSampler` with the dataloaders. During `trainer.{trainer_fn.value}()`, it is"
+                    " recommended to use `Trainer(devices=1, num_nodes=1)` to ensure each sample/batch gets evaluated"
                     " exactly once. Otherwise, multi-device settings use `DistributedSampler` that replicates"
-                    " some samples to make sure all devices have same batch size in case of uneven inputs.",
+                    " replicates some samples to make sure all devices have same batch size in case of uneven inputs.",
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:310]: Duplicated word

```suggestion
                    " some samples to make sure all devices have same batch size in case of uneven inputs.",
```

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/data_connector.py [hunk 2]
```diff
@@ -298,12 +298,16 @@ def _resolve_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional
 
             # update docs too once this is resolved
             trainer_fn = self.trainer.state.fn
-            if isinstance(sampler, DistributedSampler) and trainer_fn in (TrainerFn.VALIDATING, TrainerFn.TESTING):
+            if (
+                self.trainer.distributed_sampler_kwargs["num_replicas"] > 1
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/data_connector.py:302]: You need to get this info from the sampler directly. The user could supply one manually.

Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:302]:
```suggestion
                sampler.num_replicas > 1
```
But needs to move after the instance check.

File: tests/tests_pytorch/trainer/connectors/test_data_connector.py [hunk 3]
```diff
@@ -525,19 +525,20 @@ def test_invalid_hook_passed_in_datahook_selector():
         dh_selector.get_hook("setup")
 
 
-def test_eval_distributed_sampler_warning(tmpdir):
+@pytest.mark.parametrize("devices, warn_cm", [(1, no_warning_call), (2, pytest.warns)])
```
Comment (awaelchli) [inline tests/tests_pytorch/trainer/connectors/test_data_connector.py:528]: what does cm stand for?

File: src/pytorch_lightning/trainer/connectors/data_connector.py (hunk 2 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/data_connector.py:302]: agree! but wouldn't that create a conflict if they won't match?

File: tests/tests_pytorch/trainer/connectors/test_data_connector.py (hunk 3 above)
Comment (rohitgr7) [reply tests/tests_pytorch/trainer/connectors/test_data_connector.py]: context_manager. Updated

File: src/pytorch_lightning/trainer/connectors/data_connector.py (hunk 2 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: done!

Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/data_connector.py]: Up to the user. We have examples in the docs that show how to use a custom DistributedSampler.

**Review by awaelchli — APPROVED** [review]

**Review by otaj — APPROVED** [review]

**Review by Borda — APPROVED** [review]
